File: info/1. Алгоритмы/4. Комбинаторика/2. Генерация перестановок. Алгоритм Джонсона-Троттера/code sample/Johnson_Trotter_algorithm.py

def find_max_mobile_element(permutation, direction):
    index = -1
    for i in range(len(permutation)):
        next_element_index = i + direction[i]
        if next_element_index >= 0 and next_element_index < len(permutation):
            if permutation[i] > permutation[next_element_index]:
                if index == -1:
                    index = i
                else:
                    if permutation[i] > permutation[index]:
                        index = i
    return index


def change_direction(permutation, direction, mobile_element):
    for i in range(len(permutation)):
        if permutation[i] > mobile_element:
            direction[i] = direction[i]*(-1)


def swap_element(permutation, direction, i, j):
    permutation[i], permutation[j] = permutation[j], permutation[i]
    direction[i], direction[j] = direction[j], direction[i]


def permutation_generator(n):
    permutation = list(range(1, n+1))
    direction = [-1]*n
    print(permutation)
    mobile_element_index = find_max_mobile_element(permutation, direction)
    while mobile_element_index != -1:
        mobile_element = permutation[mobile_element_index]
        next_index = mobile_element_index + direction[mobile_element_index]
        swap_element(permutation, direction, mobile_element_index, next_index)
        change_direction(permutation, direction, mobile_element)
        print(permutation)
        mobile_element_index = find_max_mobile_element(permutation, direction)


permutation_generator(3)

File: src/algoritms/combinatorics/permutations/JohnsonTrotterAlgorithm.java
package algoritms.combinatorics.permutations;

import java.util.Arrays;

public class JohnsonTrotterAlgorithm {

    public static void main(String[] args) {
        permutationGenerator(4);
    }

    public static int findMaxMobileElement(int[] permutation, int[] direction) {
        int index = -1;
        for (int i = 0; i < permutation.length; i++) {
            int nextIndex = i + direction[i];
            if (nextIndex >= 0 && nextIndex < permutation.length) {
                if (permutation[i] > permutation[nextIndex]) {
                    if (index == -1) {
                        index = i;
                    } else {
                        if (permutation[i] > permutation[index]) {
                            index = i;
                        }
                    }
                }
            }
        }
        return index;
    }

    public static void changeDirection(int[] permutation, int[] direction, int mobileElement) {
        for (int i = 0; i < permutation.length; i++) {
            if (permutation[i] > mobileElement) {
                direction[i] = direction[i] * (-1);
            }
        }
    }

    public static void swap(int[] permutation, int[] direction, int i, int j) {
        int permutationTemp = permutation[i];
        permutation[i] = permutation[j];
        permutation[j] = permutationTemp;

        int directionTemp = direction[i];
        direction[i] = direction[j];
        direction[j] = directionTemp;
    }

    public static void permutationGenerator(int n) {
        int[] permutation = new int[n];
        int[] direction = new int[n];
        for (int i = 0; i < permutation.length; i++) {
            permutation[i] = i + 1;
            direction[i] = -1;
        }
        System.out.println(Arrays.toString(permutation));
        int mobileElementIndex = findMaxMobileElement(permutation, direction);
        while (mobileElementIndex != -1) {
            int mobileElement = permutation[mobileElementIndex];
            int nextIndex = mobileElementIndex + direction[mobileElementIndex];
            swap(permutation, direction, mobileElementIndex, nextIndex);
            changeDirection(permutation, direction, mobileElement);
            System.out.println(Arrays.toString(permutation));
            mobileElementIndex = findMaxMobileElement(permutation, direction);
        }
    }
}
